### PR TITLE
Add Plugin API for storing data in .park files

### LINKED ISSUE #14676: [Plugin] Per-save plugin storage
(I looked through existing issues and didn't find an existing issue or reference, sorry if I missed something.)

Now that the new save format is in testing and closer to being implemented, I wanted to add a feature request for adding per-save plugin storage. I'd like to be able to store settings for specific parks, and the current sharedStorage a) isn't conducive to saving per-park data, because there isn't a unique identifier (that I know of; I've seen a plugin use the park name, which isn't a constant and would clash with other saves), and even if it did, b) promotes orphaned data by separating park-specific information from the park itself and increasing the plugin.store.js file size with many saves.

This per-save storage would be in two parts:
1. "Global" storage that applies to the whole park. This could be used for defaults or settings/data that apply to the whole park.
2. Per-ride storage. This would be for data that applies to a specific ride (this _could_ be stored in global by using the ride.id, however if the ride is deleted the stored data would be useless).

As a real-world example, I have an automatic price manager that I use to set ideal prices to maximize scam- uh, _convincing_ my guests into parting with their money. One of the settings is setting ride prices to either the value price (most guests will pay and think "x is really good value"), the expensive price (most guests will pay before thinking "I'm not paying that much to go on x"), or halfway in-between. In many scenarios, typically this would just be set to expensive and left alone, since you can add ATMs and guests will continue to gleefully withdraw from their bank account. However, if you have a scenario with no ATMs and a park goal of x number of guests (read: RCT1 scenarios), you're better off using the value price so guests will stay in the park longer.

What I'd like to do is make my plugin save the settings per-save, that way I can switch between saves and have it set the proper settings for that park without accidentally blowing away the settings in another save. This would also help prevent modifying settings in new scenarios that price things differently from how I usually want them, and provide a way of keeping track of if the user has done something with the plugin for that park (for example, setting defaults for the park).

### PR DESCRIPTION
**Pull request by IntelOrca**
Resolves #14676

API is `context.parkStorage`, works similarly to `context.sharedStorage`.

### CODE REVIEW

File: src/openrct2/park/ParkFile.h
Comment (spacek531) [inline src/openrct2/park/ParkFile.h:11]: 0xA?

Comment (ZehMatt) [reply src/openrct2/park/ParkFile.h]: Why is this hexadecimal in the first place, makes more sense to have this human readable?

Comment (Broxzier) [reply src/openrct2/park/ParkFile.h]: I assumed that it's so that you can view a (corrupt) park file in a hex editor and immediately see if the version matches, though a decent hex viewer can display numeric values for the selection.

Comment (ZehMatt) [reply src/openrct2/park/ParkFile.h]: I think if you are already using a Hex Editor you have probably enough knowledge to figure that part out as well, a version number should be a decimal value, at least in the source files.

**Comment by ZehMatt** [comment]
Hm there is a slight issue with making this storage global to all plugins, they might be fighting over name collisions at some point. Can't we do this storage per plugin? If some plugin wants to access the storage of another plugin we could add an API to allow reading it but disallow modifying.

**Comment by IntelOrca** [comment]
@ZehMatt the documentation explicitly mentions that it is shared. It forces plugins to have at least one full stop in their name, so they have to have at least one namespace. Plugin writers must ensure that namespace is unique, because there will be cases where plugins may want to read and modify the same store.

**Comment by oli414** [comment]
Would it perhaps be possible to prevent someone from overwriting parkStorage as a whole by only allowing write access on its properties, not sure if that is possible.
e.g:

```javascript
parkStorage = {"oli414": null}; // not allowed
parkStorage.oli414 = null; // allowed
```

**Comment by ZehMatt** [comment]
I would try to automate this as much as possible so something like
```js
parkStorage.write("key", value);
```
would actually write to ```parkStorage.data["plugin"].key = value;``` or however its done internally.

If the plugin wants to read another plugins value
you could allow specifying the plugin name, at least for reading.
```js
var value = parkStorage.read("plugin", "key");
```
I think plugin devs shouldn't have to deal with such details, its also quite error prone giving direct access like that.

**Comment by IntelOrca** [comment]
This would differ from `sharedStorage`, surely they should be consistent?

**Comment by ZehMatt** [comment]
> This would differ from `sharedStorage`, surely they should be consistent?

I would apply this concept to any storage, it just makes a better API this way imho. The current design rather recommends using a unique entry and doesn't enforce/restrict the way it can be used.

**Comment by IntelOrca** [comment]
@ZehMatt plugins now rely on the current API. Also I personally don't have a problem with how it is currently done.

**Comment by ZehMatt** [comment]
> @ZehMatt plugins now rely on the current API. Also I personally don't have a problem with how it is currently done.

That may be the case but I would try to deprecate it. Either way I leave that up to you, I personally think that when it comes to scripting that at least the API should not be simply an after-thought.

Right now one has to document that you may have a collision if you refuse to use a new table in the root object, not the best API in my eyes.

**Comment by IntelOrca** [comment]
@ZehMatt yes but you are suggesting that two plugins can't both write to the same store which is just an unnecessary limitation.

**Comment by ZehMatt** [comment]
> Ok, how about: `context.getParkStorage()` is an alias for `context.getParkStorage(currentPluginName)`.
> 
> ```js
> var store = context.getParkStorage();
> store.set('mydata', data);
> 
> var store3 = context.getParkStorage('anotherPlugin');
> store.set('otherdata', otherdata);
> ```

Yeah that seems sensible to me, not sure I like allowing to write but I think this is a much better approach in general and plugins accessing other plugins will be probably a very rare.

**Comment by oli414** [comment]
> Ok, how about: `context.getParkStorage()` is an alias for `context.getParkStorage(currentPluginName)`.
> 
> ```js
> var store = context.getParkStorage();
> store.set('mydata', data);
> 
> var store3 = context.getParkStorage('anotherPlugin');
> store.set('otherdata', otherdata);
> ```

That would be even better. This addition of the park storage allows you to change the standard. Deprecate the way it is done on the global storage so that  that one can be done the same way.

I personally use my username as the top level namespace as well just in case:
`oli414.myPlugin.number`

File: distribution/openrct2.d.ts
Comment (ZehMatt) [inline distribution/openrct2.d.ts:190]:
```suggestion
         * Gets the storage for the current plugin if no name was specified or for the plugin of given name. Data is persisted
```
Makes this a bit more clear that this parameter should be mostly optional, at first thought it would seem like I have to pass my own plugin name there. You can probably lay out that sentence better than I can

**Review by ZehMatt — APPROVED** [review]

**Pull request merged**
